Source/WebCore/loader/ImageLoader.cpp

```diff
@@ -677,4 +678,27 @@ VisibleInViewportState ImageLoader::imageVisibleInViewport(const Document& docum
     return renderReplaced && renderReplaced->isContentLikelyVisibleInViewport() ? VisibleInViewportState::Yes : VisibleInViewportState::No;
 }
 
+bool ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive(const ImageCandidate& candidate) const
+{
+#if ENABLE(WEB_ARCHIVE) || ENABLE(MHTML)
+    if (candidate.originAttribute == ImageCandidate::SrcOrigin)
+        return false;
+
+    Ref document = element().document();
+    RefPtr loader = document->loader();
+    if (!loader || !loader->hasArchiveResourceCollection())
+        return false;
+
+    auto candidateURL = URL { element().resolveURLStringIfNeeded(candidate.string.toString()) };
+    if (loader->archiveResourceForURL(candidateURL))
+        return false;
+
+    RefPtr page = document->protectedPage();
+    return page && !page->allowsLoadFromURL(candidateURL, MainFrameMainResource::No);
```

Comment:
If `page` is null, is it right to keep this candidate and load it from the network? Should not this condition be:
```
    return !page || !page-> allowsLoadFromURL(...);
```

Reply:
Hm.. I think that in practice, it doesn't really matter — I only did it this way to preserve existing behavior in case page is null. Will change to `return !page || !page-> allowsLoadFromURL(...);`